Vector-optics birefringence reconstruction prototype

talonchandler opened the pull request:
This PR supports the [waveOrder arXiv preprint](https://arxiv.org/abs/2412.09775) and will precede a `2.2.0` release. It includes:

- all of the vector-optics functionality that supports the preprint, primarily in the `inplane_oriented_thick_pol3d_vector` model, 
- a quick-start script in `waveorder/examples/models/inplane_oriented_thick_pol3d_vector.py`,
- a new `waveorder/examples/visuals/` folder with scripts that generate Green's tensor and vector transfer function visuals,  
- a revised `/examples/README.md`, with a new table of example types, 
- a revised `README.md`, with new badges ([matching VisCy's](https://github.com/mehta-lab/VisCy/pull/197)), a new title figure, and new links to the preprint and PTI manuscript.

mattersoflight commented on waveorder/visuals/matplotlib_visuals.py:
```diff
@@ -0,0 +1,296 @@
+        },
+    )
+
+    rose_path = os.path.join(
+        os.path.dirname(__file__),
+        f"./assets/rose.png",
+    )
```
@talonchandler getting an error due to these lines:
`FileNotFoundError: [Errno 2] No such file or directory: '/hpc/mydata/shalin.mehta/code/waveorder/waveorder/visuals/./assets/rose.png'`

talonchandler replied:
Pardon me. Fixed.

talonchandler commented:
2024-12-19 TC testing notes:
- [x] automated tests passing
- [x] from-scratch installation tested manually 
- [x] all 5 `/examples/models/` tested manually
- [x] both `/examples/visuals/` scripts tested manually

mattersoflight commented on README.md:
````diff
@@ -1,57 +1,83 @@
+https://github.com/user-attachments/assets/4f9969e5-94ce-4e08-9f30-68314a905db6
 
+`waveorder` enables simulations and reconstructions of label-agnostic microscopy data as described in the following [preprint](https://arxiv.org/abs/2412.09775)
 ```bibtex
````
I suggest collapsing the bibtex entry using <details> block

mattersoflight commented on README.md:
````diff
@@ -1,57 +1,83 @@
+The acquisition, calibration, background correction, reconstruction, and applications of PTI are described in the following [paper](https://doi.org/10.1101/2020.12.15.422951):
 
-In addition to label-free reconstruction algorithms, `waveorder` also implements widefield fluorescence and fluorescence polarization reconstruction algorithms for correlative label-free and fluorescence imaging.
+```bibtex
````
I suggest wrapping the bibtex entry in <details> block

ziw-liu commented on README.md:
````diff
@@ -68,6 +94,13 @@ python
 
+(Optional) Install napari, clone the repository, and run an example script:
+```sh
+pip install napari[all]
````
We can potentially put napari and jupyter in optional dependencies e.g. `pip install "waveorder[examples]".

talonchandler replied:
Great idea!

I just updated `pyproject.toml` so that `pip install "waveorder[examples]"` installs napari and jupyter, tested a fresh install, and updated both `README`s.

ziw-liu commented on waveorder/visuals/jupyter_visuals.py:
```diff
@@ -176,8 +172,7 @@ def image_stack_viewer_fast(
 
-    im_wgt = Image(
-        value=im_dict[0],
+    im_wgt = Image(        value=im_dict[0],
```
Unintended diff?

talonchandler replied:
Good eye. Fixed.

ziw-liu commented on waveorder/visuals/matplotlib_visuals.py:
These functions are quite long and could need some type hints and docstrings.

talonchandler replied:
Agreed! Added.

ziw-liu commented on waveorder/visuals/napari_visuals.py:
```diff
@@ -0,0 +1,75 @@
+
+
+def add_transfer_function_to_viewer(
+    viewer,  # napari viewer
```
Type hint can be guarded in `typing.TYPE_CHECKING` or in strings.

talonchandler replied:
Thank you!

talonchandler replied:
Hmmm I couldn't figure out how to get `TYPE_CHECKING` to work...even after comparing with what you've been doing in recOrder. 

Changed to strings.

ziw-liu replied:
That needs `from __future__ import annotations` to work. String is equivalent though.

ziw-liu commented on waveorder/models/isotropic_thin_3d.py:
```diff
@@ -169,7 +168,7 @@ def visualize_transfer_function(
             scale=(1, 1, 1),
         )
-    viewer.dims.order = (0, 1, 2)
+    viewer.dims.order = (2, 0, 1)
```
Why is this one not using the util function?

talonchandler replied:
Great question. This transfer function is our only 3d-to-2d transfer function, so it stands alone for now. I've also added a comment addressing this.

ziw-liu commented on waveorder/util.py:
```diff
@@ -2239,3 +2242,67 @@ def orientation_3D_continuity_map(
+    return sigma
+
+
+# torch.allclose(
+#     torch.abs(torch.einsum("kij,lji->kl", s, s) - torch.eye(4)),
+#     torch.zeros((4, 4)),
+#     atol=1e-5,
+# )
```
```suggestion
```

ziw-liu commented on waveorder/util.py:
```diff
@@ -2239,3 +2242,67 @@ def orientation_3D_continuity_map(
+        ], dtype=torch.complex64
+    )
+
+    # torch.allclose(
+
+
+#     torch.abs(torch.einsum("kij,lji->kl", Y, Y) - torch.eye(9)),
+#     torch.zeros((9, 9)),
+#     atol=1e-5,
+# )
```
```suggestion
```
move to test?

talonchandler replied:
Thanks for flagging! I meant to do this and forgot. Moved to tests.

ziw-liu commented on waveorder/visuals/assets/gellman/0.png:
Why do these PNGs need to be in the library (and hence in the PyPI wheel)?

talonchandler replied:
I agree. Removed.

ziw-liu commented on waveorder/waveorder_reconstructor.py:
We should put a deprecation warning here.

talonchandler replied:
Good idea. Done.

mattersoflight commented on examples/README.md:
```diff
@@ -1,7 +1,10 @@
+| Folder      | Requires                   | Description                                                                                           |
+|------------------|----------------------------|-------------------------------------------------------------------------------------------------------|
+| `models/`        | `waveorder`, `napari`      | Demonstrates the latest functionality of `waveorder` through simulations and reconstructions using various models. |
+| `maintenance/`   | `waveorder`                | Showcases functionality planned for `models/` folder; scripts are maintained with automated tests.               |
```
Do you mean, "Examples of computational imaging methods enabled by functionality of `waveorder`; these scripts are maintained with automated tests"

talonchandler replied:
That's correct...fixed. Thanks.

ziw-liu commented:
Overall I think at least the core public API needs some documentation, even just type hints will help with future maintenance.

mattersoflight reviewed (approved):
The PR looks great to me! I reviewed two READMEs (main and examples) and the new example: `inplane_oriented_thick_pol3d_vector.py`. 

Minor input on example: label the spatial and channel axes in napari viewer, e.g., first axis when data is displayed is `z` and second axis is `c`. For transfer functions, the axes seem to be `y`, `d` (for data), `s` (for object).

talonchandler commented:
Thank you both!

@mattersoflight I've addressed your comments and added axis labels (great idea). 

@ziw-liu I've addressed your comments and added docstrings and type hints in the most-used APIs throughout `waveorder.models` and `waveorder.visuals`.
